### DISCUSSION TITLE
GraphQL error: Cannot read property 'code' of null when save page

### DISCUSSION THREAD

**Discussion by luanknebel** [requarks/wiki#3703] [Error / Bug Report]
GraphQL error: Cannot read property 'code' of null

**Comment by AuspeXeu** [comment]
I image it would be much easier to help you if you provided any context information...

**Comment by cyruscuenca** [comment]
I'm having the same issue. I'll expand with context:

I installed WikiJS using this guide: https://docs.requarks.io/install/linux

The VPS is running Ubuntu 18.04, and WikiJS is being run with admin privileges with pm2. When I SSH into the VPS, kill it, and restart it, the app works. However, as soon as I terminate the SSH connection, the app breaks, and anytime I try to save a page, I get:

``Cannot read property 'code' of null when save page``

**Comment by cyruscuenca** [comment]
Also, when I run the app as a service as specified in the docs, the error changes to:

``Network error: JSON.parse: unexpected character at line 1 column 1 of the JSON data ``

**Comment by coinfastman** [comment]
Good afternoon after the transition to the new version 2.0.12, also an error. The update was performed according to the instructions in your documentation. But after the update I can not edit the page. I get a **GraphQL error: Cannot read property 'code' of null**. Installed wikijs on ubuntu 18.04 MySQL is used as the database.
![Снимок экрана от 2019-12-30 18-24-12](https://user-images.githubusercontent.com/33808421/71588396-5a34f280-2b32-11ea-8dd8-1f392995a678.png)

**Comment by ai-wiki** [comment]
I'm getting a similar issue in Wiki.js 2.0.12 with postgreSQL after migrating a wiki from a Heroku instance to a Linux server. The wiki pages were backed up to a git repository, then imported into the newly created wiki.

I can create and edit pages normally in the new wiki, but pages imported from the old wiki fail to load with "Error while fetching page. Duplicate entry detected. Reload the page to try again", and attempting to edit them result in the "Cannot read property 'code' of null" error when saving.

**Comment by majidalavizadeh** [comment]
I have this problem too, I have imported the pages from Github repository and I cannot edit the pages.

Version: 2.0.12

**Comment by coinfastman** [comment]
Restarting wikijs helps, but the effect is not long.

**Comment by NGPixel** [comment]
Unfortunately `Cannot read property 'code' of null` is a very generic error. The next update (2.1) will come with better error details to better pinpoint the root issue.

**Comment by Coconop** [comment]
Same issue:

**OS:** Debian GNU/Linux 9.11 (stretch)
**Wiki.js:** 2.0.12
**Database:** MySQL 10.4.11-MariaDB

I could not find a consistent way of reproducing it, but at some point when I create a page and save it, I got the `GraphQL error: Cannot read property 'code' of null` when I save the page.
The page is actually created and I can see it from the administration dashboard, however I cannot view it and got the error `Error while fetching page. Duplicate entry detected. Reload the page to try again`.
It is still possible to view the sources or edit it. 
If I try to remove it from the dashboard I have `An unexpected error occurred`. And the "Delete" button is spinning forever. Clicking out of the remove pop-up closes it but the page is still listed.
However, if the node server is relaunched, the removed pages are not listed anymore.

**EDIT** Bug is still present with version 2.1.113

**Comment by levistepanov** [comment]
Same issue

**OS:**: Centos 7
**wiki.js**: 2.0.12
**Database**: Amazon RDS db.t3.small mysql

**Error**: "GraphQL error: Cannot read property 'code' of null"

happens when trying to edit or create new pages. Was working after initial installation, then stopped after a couple days.

**Comment by coconut123** [comment]
Same issue here : 

**OS :** Debian 10 
**wiki.js** : 2.1.113
**Database** : postgresql 11.7 

happens when trying to edit or create new pages. Stop working afer update from 2.0.12 to 2.1.113

**Comment by tostr7191** [comment]
I am having the same issue.

OS: Alpine Linux 3.11.3
wiki.js: 2.1.113
DB: psql 12.2
node: 12.15.0

The wikijs is behind a nginx proxy, but it does not matter if I access it directly on the server.

**Comment by roman-akhremko** [comment]
Problem is still actual after upgrade to 2.2.51

OS: Ubuntu 18.04.3 (LTS) x64
wiki.js: 2.2.51
DB: PostgreSQL 10.12
node: 12.16.1

**Comment by vincent-polite** [comment]
What is the recommended workaround for this issue?  Restart and pray?  This just started happening with a page I am editing.

**Comment by hmbravo** [comment]
I have the same issue, when you plan solve this problem?

**Comment by WalterMontalvo** [comment]
After a fresh install of 2.3.77, I get this error when saving edits on an existing page.
"GraphQL error: Cannot read property 'code' of null"

After stop/start of the server, the error goes away.

Yeah, lots of detail, because you never know what can be the clue that solves the mystery.

Definitely having more diagnostic info would be good. Perhaps we could add a "diagnostic level" setting to specify how granular of diagnostic output we want.


Navigation path:
-----------------
Click on a Markdown page created and edited the day before.
Click on Pencil icon, bottom right, to edit page.
Modify a single character in markdown tag.
Click Save
Get popup with error "GraphQL error: Cannot read property 'code' of null"


Chronology:
-------------
1. Installed wiki.js 2.3.77.  Modified config.yml to connect to PostgreSQL.  Look at Server details below for specific version numbers.
2. Started node server
3. On initial setup, instead of creating the home page, I created another page (Test-Styles) under home/ subdirectory.  So /home/Test-Styles
4. Created home page.  On save, it raised an error, but it created the page.  Subsequent edits on the home page did not raise any errors.
Now I have
/ home - as the home page
/ home / Test-Styles - initial page I created
(Is this weird or what?)
5. Stopped and started server a few times and created and edited a few test pages.
6. Setup Git storage.  Used Amazon AWS CodeCommit Git repository.  Amazon's Git does not accept credentials in the URL, so had to modify the storage.js code to not embed the credentials.  Manually saved the credentials via UNIX Git commands.
7. Tested bidirectional changes.  So changed files via wiki.js, it propagates to Amazon Git.  Changes in Amazon git, propagate back to wiki.js.  Bidirectional syncing is setup for every 5 minutes.
8. Edit a few more files.
9. Keep server running overnight.
10. 11 hours later, when trying to save modifications to an existing Markdown page, it raises the error
"GraphQL error: Cannot read property 'code' of null"
11. After stop/start of the server, the error goes away.

Server details:
--------------
OS: Linux - CentOS (core) 7.7.1908 x64
Current version: 2.3.77
Initial version on install: 2.3.77
DB: PostgreSQL 10.11
Node: 14.1.0

**Comment by showvin** [comment]
I have the same issue, my environment description is as follows:

WiKi: 2.4.105
DB : PostgreSQL 10.8 (Ubuntu 10.8-1.pgdg14.04+1)
OS : Ubuntu 16.04.6 LTS

**Comment by aurora** [comment]
Same issue here, first time i got this error after using wikijs quite a while now. Error ocurred when i wanted to save changes after editing a page. The page content is actually saved, but the view is not updated. Works again after restarting node server.

Wiki: 2.3.81
DB: PostgreSQL 10.12
OS: Ubuntu 18.04 LTS

**Comment by showvin** [comment]
I found steps that can always reproduce this problem

1. login the Ubuntu server with SSH（Putty/SecureCRT）
2. login as root
3. node server &
4. disconnect SSH connection

The problem can be reproduced in a few hours.
If I keep the SSH connection, everything works ok.

**Comment by SebastienLegrand** [comment]
If I launched the wiki with cron, I can disconnect from the SSH connection and it's still working.

But When I try to stop it, He doesn't stop and I need to kill all node processes

**Comment by dctopspin** [comment]
Same issue. Brand new install on AWS with docker. Can't even save the first page

**Comment by mathieufrh** [comment]
I use tmux as a workaround. But `node server & disown` is not working.

**Comment by slecuona** [comment]
I resolved this issue by switching pm2 exec_mode from **clustered** to **fork** (default).
Version 1 of WikiJS needs exec in clustered mode, but seems that version 2, not. I suggest add this clarification in the documentation.

**Comment by ghost** [comment]
Same issue. I'm use the command `node server &`

Wiki.js
Current Version
2.5.159
Host Information
Operating System: Linux - Debian (buster) 10 x64
Hostname: SRVWIKI
CPU: Cores 2
Total RAM 3.83 GB
Working Directory
/etc/opt/wiki
Configuration File
/etc/opt/wiki/config.yml
Node.js 12.18.3
PostgreSQL 11.7

**Comment by Im-Fran** [comment]
Hi, I got the same issue, I was editing a page and now I can't save it, would help you if I share my telemetry client id?

System info:
- WikiJS Version: 2.5.170
- OS: Linux - Ubuntu (bionic) 18.04 x64
- CPU Cores: 2
- RAM: 3.76GB
- MySQL Version: 8.0.20
- NodeJS Version:  12.16.2

**Comment by Im-Fran** [comment]
@NGPixel Sorry for the ping but this issue is not letting me to edit any page, and this has been open since 2019, we are in 2020, I don't know if you could fix this or your team could do it but this is a bug that is not letting me use the most important part of the website, I can't edit any page because of this error. I will attach a video of the problem.

https://user-images.githubusercontent.com/30329003/104363321-c55e2680-54f3-11eb-912c-92e2d8704e97.mov

**Comment by NGPixel** [comment]
@Im-Fran This is a very generic error and not something that can be easily fixed. Try disabling any storage modules you enabled, disable rendering modules you manually enabled, etc.

**Comment by Im-Fran** [comment]
Ok, I will try it, thanks for the answer.

**Comment by Im-Fran** [comment]
> @Im-Fran This is a very generic error and not something that can be easily fixed. Try disabling any storage modules you enabled, disable rendering modules you manually enabled, etc.

I disabled them and still the same error, you want me to share my telemetry client id? Maybe that would help you to identify the problem... Since it's the only problem I have

**Comment by Im-Fran** [comment]
Hey @NGPixel sorry for the ping again, so, since this doesn't works, there is no other solution for this? I still think that it might be helpful if I share with you my telemetry client id so you can check the logs or something like that to see why is this hapenning, and I'm seeing that I'm not the only one with this issue, this issue is blocking the main functionality of the website, this started hapenning after the last update, and I don't want to downgrade my installation, so is there any solution, something I can help you with to find a fix for this?

**Comment by Im-Fran** [comment]
Ok, I just got this issue fixed, I reinstalled the WikiJS instance and that worked. Hope it helps anyone with the same issue

**Comment by or1is1** [comment]
@NGPixel 

I solved the problem.

from the official guide (=https://docs.requarks.io/install/linux), 'Consider creating a dedicated user for Wiki.js here'
If we just follow that guide(it means didn't create any user for wiki.js), the proble(graphql  error: ~) has occured.

So, we have to create the user for wiki.js like named 'wiki', It solved.

Follow below command.
```bash
sudo groupadd --system wiki
sudo useradd -s /sbin/nologin --system -g wiki wiki
sudo nano /etc/systemd/system/wiki.service
```
```bash
[Unit]
Description=Wiki.js
After=network.target

[Service]
Type=simple
ExecStart=/usr/bin/node server
Restart=always
User=wiki
Environment=NODE_ENV=production
WorkingDirectory=/srv/wiki

[Install]
WantedBy=multi-user.target
```
```bash
sudo chown -R wiki:wiki /srv/wiki
sudo systemctl daemon-reload
sudo systemctl enable --now wiki.service
```

I found the key frome here -> https://computingforgeeks.com/how-to-install-wiki-js-on-centos-8/

**Comment by missytake** [comment]
Could this maybe be related to this issue? https://github.com/Requarks/wiki/discussions/3340 There is more debug information in it.
